File: router/src/protocol/cancelReason.ts
export const REQUEST_CANCEL_REASONS = [
  'timeout',
  'caller_cancel',
  'runtime_disconnect',
  'gateway_disconnect',
  'drain',
  'retire',
  'client_disconnect',
  'router_shutdown',
  'backpressure',
  'deadline_exceeded',
  'protocol_error',
  'stream_dropped'
] as const;

export type RequestCancelReason = (typeof REQUEST_CANCEL_REASONS)[number];

export const CONTRACT_H_REQUEST_CANCEL_SITUATIONS = [
  'caller_abort',
  'client_disconnect',
  'timeout',
  'deadline_exceeded',
  'backpressure',
  'protocol_error',
  'stream_dropped',
  'runtime_disconnect',
  'router_shutdown'
] as const;

export type RequestCancelSituation = (typeof CONTRACT_H_REQUEST_CANCEL_SITUATIONS)[number];

export const REQUEST_CANCEL_SITUATION = {
  callerAbort: 'caller_abort',
  clientDisconnect: 'client_disconnect',
  timeout: 'timeout',
  deadlineExceeded: 'deadline_exceeded',
  backpressure: 'backpressure',
  protocolError: 'protocol_error',
  streamDropped: 'stream_dropped',
  runtimeDisconnect: 'runtime_disconnect',
  routerShutdown: 'router_shutdown'
} as const satisfies Record<string, RequestCancelSituation>;

export const REQUEST_CANCEL_REASON_BY_SITUATION = {
  caller_abort: 'caller_cancel',
  client_disconnect: 'client_disconnect',
  timeout: 'timeout',
  deadline_exceeded: 'deadline_exceeded',
  backpressure: 'backpressure',
  protocol_error: 'protocol_error',
  stream_dropped: 'stream_dropped',
  runtime_disconnect: 'runtime_disconnect',
  router_shutdown: 'router_shutdown'
} as const satisfies Record<RequestCancelSituation, RequestCancelReason>;

export interface RequestCancelReasonMapping {
  internalReason: string;
  wireReason: RequestCancelReason;
}

const KNOWN_REQUEST_CANCEL_REASONS = new Set<string>(REQUEST_CANCEL_REASONS);

const INTERNAL_REASON_FALLBACKS = {
  caller_abort: REQUEST_CANCEL_REASON_BY_SITUATION.caller_abort,
  unexpected_stream_response: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  unexpected_control_response: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  response_channel_closed: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  duplicate_response_start: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  chunk_before_start: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  chunk_seq_mismatch: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  chunk_decode_error: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  stream_end_payload: REQUEST_CANCEL_REASON_BY_SITUATION.protocol_error,
  stream_cancelled: REQUEST_CANCEL_REASON_BY_SITUATION.stream_dropped
} as const satisfies Record<string, RequestCancelReason>;

export function isRequestCancelReason(reason: string): reason is RequestCancelReason {
  return KNOWN_REQUEST_CANCEL_REASONS.has(reason);
}

export function requestCancelReasonForSituation(
  situation: RequestCancelSituation
): RequestCancelReason {
  return REQUEST_CANCEL_REASON_BY_SITUATION[situation];
}

export function mapInternalRequestCancelReason(internalReason: string): RequestCancelReasonMapping {
  if (isRequestCancelReason(internalReason)) {
    return { internalReason, wireReason: internalReason };
  }
  return {
    internalReason,
    wireReason:
      INTERNAL_REASON_FALLBACKS[internalReason as keyof typeof INTERNAL_REASON_FALLBACKS] ??
      REQUEST_CANCEL_REASON_BY_SITUATION.caller_abort
  };
}

File: runtime/transport/src/cancel_reason.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RequestCancelReason {
    Timeout,
    CallerCancel,
    RuntimeDisconnect,
    GatewayDisconnect,
    Drain,
    Retire,
    ClientDisconnect,
    RouterShutdown,
    Backpressure,
    DeadlineExceeded,
    ProtocolError,
    StreamDropped,
}

impl RequestCancelReason {
    pub const CONTRACT_H: [RequestCancelReason; 9] = [
        RequestCancelReason::CallerCancel,
        RequestCancelReason::ClientDisconnect,
        RequestCancelReason::Timeout,
        RequestCancelReason::DeadlineExceeded,
        RequestCancelReason::Backpressure,
        RequestCancelReason::ProtocolError,
        RequestCancelReason::StreamDropped,
        RequestCancelReason::RuntimeDisconnect,
        RequestCancelReason::RouterShutdown,
    ];

    /// Whether this reason is a frozen `CONTRACT_H` wire value
    /// (C-model-request §4). The codec rejects any other reason.
    pub const fn is_contract_h(self) -> bool {
        matches!(
            self,
            RequestCancelReason::CallerCancel
                | RequestCancelReason::ClientDisconnect
                | RequestCancelReason::Timeout
                | RequestCancelReason::DeadlineExceeded
                | RequestCancelReason::Backpressure
                | RequestCancelReason::ProtocolError
                | RequestCancelReason::StreamDropped
                | RequestCancelReason::RuntimeDisconnect
                | RequestCancelReason::RouterShutdown
        )
    }

    /// Parses a wire reason and rejects values outside the frozen
    /// `CONTRACT_H` vocabulary (legacy internal reasons such as
    /// `gateway_disconnect` / `drain` / `retire` are not legal wire values).
    pub fn from_contract_h_wire(reason: &str) -> Option<Self> {
        let parsed = Self::from_wire(reason)?;
        parsed.is_contract_h().then_some(parsed)
    }

    pub fn as_str(self) -> &'static str {
        match self {
            RequestCancelReason::Timeout => "timeout",
            RequestCancelReason::CallerCancel => "caller_cancel",
            RequestCancelReason::RuntimeDisconnect => "runtime_disconnect",
            RequestCancelReason::GatewayDisconnect => "gateway_disconnect",
            RequestCancelReason::Drain => "drain",
            RequestCancelReason::Retire => "retire",
            RequestCancelReason::ClientDisconnect => "client_disconnect",
            RequestCancelReason::RouterShutdown => "router_shutdown",
            RequestCancelReason::Backpressure => "backpressure",
            RequestCancelReason::DeadlineExceeded => "deadline_exceeded",
            RequestCancelReason::ProtocolError => "protocol_error",
            RequestCancelReason::StreamDropped => "stream_dropped",
        }
    }

    pub fn from_wire(reason: &str) -> Option<Self> {
        match reason {
            "timeout" => Some(RequestCancelReason::Timeout),
            "caller_cancel" => Some(RequestCancelReason::CallerCancel),
            "runtime_disconnect" => Some(RequestCancelReason::RuntimeDisconnect),
            "gateway_disconnect" => Some(RequestCancelReason::GatewayDisconnect),
            "drain" => Some(RequestCancelReason::Drain),
            "retire" => Some(RequestCancelReason::Retire),
            "client_disconnect" => Some(RequestCancelReason::ClientDisconnect),
            "router_shutdown" => Some(RequestCancelReason::RouterShutdown),
            "backpressure" => Some(RequestCancelReason::Backpressure),
            "deadline_exceeded" => Some(RequestCancelReason::DeadlineExceeded),
            "protocol_error" => Some(RequestCancelReason::ProtocolError),
            "stream_dropped" => Some(RequestCancelReason::StreamDropped),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum RequestCancelSituation {
    CallerAbort,
    ClientDisconnect,
    Timeout,
    DeadlineExceeded,
    Backpressure,
    ProtocolError,
    StreamDropped,
    RuntimeDisconnect,
    RouterShutdown,
}

impl RequestCancelSituation {
    pub const CONTRACT_H: [RequestCancelSituation; 9] = [
        RequestCancelSituation::CallerAbort,
        RequestCancelSituation::ClientDisconnect,
        RequestCancelSituation::Timeout,
        RequestCancelSituation::DeadlineExceeded,
        RequestCancelSituation::Backpressure,
        RequestCancelSituation::ProtocolError,
        RequestCancelSituation::StreamDropped,
        RequestCancelSituation::RuntimeDisconnect,
        RequestCancelSituation::RouterShutdown,
    ];
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RequestCancelReasonMapping<'a> {
    pub internal_reason: &'a str,
    pub wire_reason: RequestCancelReason,
}

pub fn request_cancel_reason_for_situation(
    situation: RequestCancelSituation,
) -> RequestCancelReason {
    match situation {
        RequestCancelSituation::CallerAbort => RequestCancelReason::CallerCancel,
        RequestCancelSituation::ClientDisconnect => RequestCancelReason::ClientDisconnect,
        RequestCancelSituation::Timeout => RequestCancelReason::Timeout,
        RequestCancelSituation::DeadlineExceeded => RequestCancelReason::DeadlineExceeded,
        RequestCancelSituation::Backpressure => RequestCancelReason::Backpressure,
        RequestCancelSituation::ProtocolError => RequestCancelReason::ProtocolError,
        RequestCancelSituation::StreamDropped => RequestCancelReason::StreamDropped,
        RequestCancelSituation::RuntimeDisconnect => RequestCancelReason::RuntimeDisconnect,
        RequestCancelSituation::RouterShutdown => RequestCancelReason::RouterShutdown,
    }
}

pub fn map_internal_request_cancel_reason(internal_reason: &str) -> RequestCancelReasonMapping<'_> {
    let wire_reason =
        RequestCancelReason::from_wire(internal_reason).unwrap_or(match internal_reason {
            "caller_abort" => RequestCancelReason::CallerCancel,
            "unexpected_stream_response"
            | "unexpected_control_response"
            | "response_channel_closed"
            | "duplicate_response_start"
            | "chunk_before_start"
            | "chunk_seq_mismatch"
            | "chunk_decode_error"
            | "stream_end_payload" => RequestCancelReason::ProtocolError,
            "stream_cancelled" => RequestCancelReason::StreamDropped,
            _ => RequestCancelReason::CallerCancel,
        });

    RequestCancelReasonMapping {
        internal_reason,
        wire_reason,
    }
}

pub fn request_cancel_wire_reason_for_internal(internal_reason: &str) -> &'static str {
    map_internal_request_cancel_reason(internal_reason)
        .wire_reason
        .as_str()
}

#[cfg(test)]
mod tests;
